instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/ClosuresTest.kt

```diff
@@ -0,0 +1,238 @@
+
+        mapboxNavigation.routeRefreshController.requestImmediateRouteRefresh()
+
+        val firstRefresh = mapboxNavigation.routesUpdates()
+            .filter { it.reason == RoutesExtra.ROUTES_UPDATE_REASON_REFRESH }
+            .first()
+
+        val routeProgressAfterRefresh = mapboxNavigation.routeProgressUpdates()
+            .filter { it.currentState == RouteProgressState.TRACKING }
+            .take(3).toList().last()
+
+        assertFalse(routeProgressAfterRefresh.hasUnexpectedUpcomingClosures())
+
+        mockWebServerRule.requestHandlers.removeLast()
+        mockWebServerRule.requestHandlers.add(
+            FailByRequestMockRequestHandler(
+                MockDirectionsRefreshHandler(
+                    "route_response_route_refresh",
+                    // [1, 3] closure on leg#0 - closure moved
+                    readRawFileText(context, R.raw.route_response_route_refresh_annotations),
+                    acceptedGeometryIndex = 0
+                )
+            )
+        )
+        mapboxNavigation.routeRefreshController.requestImmediateRouteRefresh()
+        mapboxNavigation.routesUpdates()
+            .filter { it.reason == RoutesExtra.ROUTES_UPDATE_REASON_REFRESH && it !== firstRefresh }
+            .first()
+
+        val routeProgressAfterRefreshWithClosure = mapboxNavigation.routeProgressUpdates()
+            .filter { it.currentState == RouteProgressState.TRACKING }
+            .take(3).toList().last()
+
+        assertTrue(routeProgressAfterRefreshWithClosure.hasUnexpectedUpcomingClosures())
+    }
+
+    private fun setUpRequestHandler(file: Int) {
+        mockWebServerRule.requestHandlers.clear()
+        mockWebServerRule.requestHandlers.add(
+            MockDirectionsRequestHandler(
+                profile = "driving-traffic",
+                jsonResponse = readRawFileText(context, file),
+                expectedCoordinates = null,
+                relaxedExpectedCoordinates = true
+            )
+        )
+    }
+
+    private suspend fun requestRoutes(coordinates: List<Point>): List<NavigationRoute> {
+        return mapboxNavigation.requestRoutes(
+            RouteOptions.builder().applyDefaultNavigationOptions()
+                .profile(DirectionsCriteria.PROFILE_DRIVING_TRAFFIC)
+                .alternatives(false)
+                .enableRefresh(true)
+                .coordinatesList(coordinates)
+                .baseUrl(mockWebServerRule.baseUrl)
+                .build()
+        ).getSuccessfulResultOrThrowException().routes
+    }
+
+    private fun stayOnPosition(
+        point: Point,
+        bearing: Float = 0f
+    ) {
+        mockLocationReplayerRule.loopUpdate(
+            mockLocationUpdatesRule.generateLocationUpdate {
+                this.latitude = point.latitude()
+                this.longitude = point.longitude()
+                this.bearing = bearing
+            },
+            times = 120
+        )
+    }
```

Comment:
you can reuse existing `stayOnPosition { }` that creates a scope for location updates. It's probably a taste preference, so I'm just letting you know that this feature already exists

Reply:
Yeah, I've seen it. Probably really it's a taste preference: the code gets moved to the right with the block approach.